### PR TITLE
Start adding CT frontend. Create CT v1 handlers and hook up to RPC client

### PR DESCRIPTION
**Pull request by Martin2112**
Includes some basic request validation. Also add tests and mock log RPC client generated by mockery.

### CODE REVIEW

File: ct_fe/ct_handlers.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+const (
+	// All RFC6962 requests start with this base path
+	ctV1BasePath string = "/ct/v1/"
+	// You'd think these would be defined in some library but if so I haven't found it yet
+	httpMethodPost = "POST"
+	httpMethodGet  = "GET"
+)
+
+type CtRequestHandlers struct {
```
Comment (AlCutter) [inline ct_fe/ct_handlers.go]: Ct -> CT

Comment (AlCutter) [inline ct_fe/ct_handlers.go]: godoc comment missing

File: ct_fe/ct_handlers.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	rpcClient trillian.TrillianLogClient
+}
+
+func NewCtRequestHandlers(rpcClient trillian.TrillianLogClient) *CtRequestHandlers {
```
Comment (AlCutter) [inline ct_fe/ct_handlers.go]: NewCT... + godoc :)

File: ct_fe/ct_handlers.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+		if err != nil {
+			return
+		}
+
+		http.Error(w, http.StatusText(http.StatusNotImplemented), http.StatusNotImplemented)
+	}
+}
+
+func wrappedGetSthHandler(rpcClient trillian.TrillianLogClient) http.HandlerFunc {
```
Comment (AlCutter) [inline ct_fe/ct_handlers.go]: ...GetSTH...

File: ct_fe/ct_handlers.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !enforceMethod(w, r, httpMethodGet) {
+			return
+		}
+
+		http.Error(w, http.StatusText(http.StatusNotImplemented), http.StatusNotImplemented)
+	}
+}
+
+func wrappedGetSthConsistencyHandler(rpcClient trillian.TrillianLogClient) http.HandlerFunc {
```
Comment (AlCutter) [inline ct_fe/ct_handlers.go]: ...STH...

File: ct_fe/ct_handlers_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
+package ct_fe
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/google/trillian"
```
Comment (AlCutter) [inline ct_fe/ct_handlers_test.go]: import ordering

File: ct_fe/ct_server/main.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"flag"
+	"net/http"
+	"strconv"
+
+	"github.com/google/trillian/ct_fe"
+	"github.com/golang/glog"
+	"google.golang.org/grpc"
```
Comment (AlCutter) [inline ct_fe/ct_server/main.go]: import ordering

File: ct_fe/ct_server/main.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	// get started. Uses a blocking connection so we don't start serving before we're connected
+	// to backend.
+	conn, err := grpc.Dial(*rpcBackendFlag, grpc.WithInsecure(), grpc.WithBlock())
+
+	if err != nil {
+		glog.Fatalf("Could not connect to rpc server: %v", err)
+	}
+
+	defer conn.Close()
+  client := trillian.NewTrillianLogClient(conn)
```
Comment (AlCutter) [inline ct_fe/ct_server/main.go]: weird indentation

File: mock_log_client.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+import (
+	_ "crypto/sha256"
+
+	"github.com/stretchr/testify/mock"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// TrillianLogClient is an autogenerated mock type for the TrillianLogClient type
```
Comment (AlCutter) [inline mock_log_client.go]: "MockTrillianLogClient is ..."

File: ct_fe/ct_handlers.go (hunk 1 above)
Comment (Martin2112) [inline ct_fe/ct_handlers.go]: Done x 2.

File: ct_fe/ct_handlers.go (hunk 2 above)
Comment (Martin2112) [inline ct_fe/ct_handlers.go]: Done.

File: ct_fe/ct_handlers.go (hunk 4 above)
Comment (Martin2112) [inline ct_fe/ct_handlers.go]: Done.

File: ct_fe/ct_handlers.go (hunk 3 above)
Comment (Martin2112) [inline ct_fe/ct_handlers.go]: Done.

File: ct_fe/ct_handlers_test.go (hunk 5 above)
Comment (Martin2112) [inline ct_fe/ct_handlers_test.go]: Done.

File: ct_fe/ct_server/main.go (hunk 6 above)
Comment (Martin2112) [inline ct_fe/ct_server/main.go]: Done.

File: ct_fe/ct_server/main.go (hunk 7 above)
Comment (Martin2112) [inline ct_fe/ct_server/main.go]: Ran gofmt. Try this.

File: mock_log_client.go (hunk 8 above)
Comment (Martin2112) [inline mock_log_client.go]: Done.

File: ct_fe/ct_server/main.go [hunk 9]
```diff
@@ -27,11 +27,11 @@ func main() {
...
-  client := trillian.NewTrillianLogClient(conn)
+	client := trillian.NewTrillianLogClient(conn)
 
 	// Create and register the handlers using the RPC client we just set up
-	handlers := ct_fe.NewCtRequestHandlers(client)
+	handlers := ct_fe.NewCTRequestHandlers(client)
 	handlers.RegisterCTHandlers()
 
-	glog.Warningf("Server exited: %v", http.ListenAndServe("localhost:" + strconv.Itoa(*serverPortFlag), nil))
+	glog.Warningf("Server exited: %v", http.ListenAndServe("localhost:"+strconv.Itoa(*serverPortFlag), nil))
```
Comment (AlCutter) [inline ct_fe/ct_server/main.go]: could also do `fmt.Sprintf("localhost:%d", *serverPortFlag)`

Comment (Martin2112) [inline ct_fe/ct_server/main.go]: Might leave this and fix later. This main is only a strawman to see it running.

Comment (Martin2112) [inline ct_fe/ct_server/main.go]: Oh well, I've done it anyway.

**Pull request merged**